src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/cursor/CursorResultExecutor.java

```diff
@@ -0,0 +1,146 @@
+         * All cursor's are of the form <cursorType>:<base64 encoded cursor>
+         * The serialized form before encoding is upto Cursor implementation
+         */
+        String[] splittedCursor = cursorId.split(":");
+
+        if (splittedCursor.length!=2) {
+            throw new VerificationException("Not able to parse invalid cursor");
+        }
+
+        String type = splittedCursor[0];
+        CursorType cursorType = CursorType.getById(type);
+
+        if (cursorType!=CursorType.NULL) {
+            switch(cursorType) {
+                case DEFAULT:
+                    DefaultCursor defaultCursor = DefaultCursor.from(splittedCursor[1]);
+                    return handleDefaultCursorRequest(client, defaultCursor);
+                case AGGREGATION:
+                case JOIN:
+                default: throw new VerificationException("Unsupported cursor");
+            }
+        }
+
+        throw new VerificationException("Not able to parse invalid cursor");
+    }
+
+    private String handleDefaultCursorRequest(Client client, DefaultCursor cursor) {
+        String previousScrollId = cursor.getScrollId();
+        LocalClusterState clusterState = LocalClusterState.state();
+        TimeValue scrollTimeout = clusterState.getSettingValue(CURSOR_KEEPALIVE);
+        SearchResponse scrollResponse = client.prepareSearchScroll(previousScrollId).setScroll(scrollTimeout).get();
+        SearchHits searchHits = scrollResponse.getHits();
+        SearchHit[] searchHitArray = searchHits.getHits();
+        String newScrollId = scrollResponse.getScrollId();
+
+        int rowsLeft = (int) cursor.getRowsLeft();
+        int fetch = cursor.getFetchSize();
+
+        if (rowsLeft < fetch && rowsLeft < searchHitArray.length) {
+            /**
+             * This condition implies we are on the last page, and we might need to truncate the result from SearchHit[]
+             * Avoid truncating in following two scenarios
+             * 1. number of rows to be sent equals fetchSize
+             * 2. size of SearchHit[] is already less that rows that needs to be sent
+             *
+             * Else truncate to desired number of rows
+             */
+            SearchHit[] newSearchHits = Arrays.copyOf(searchHitArray, rowsLeft);
+            searchHits = new SearchHits(newSearchHits, searchHits.getTotalHits(), searchHits.getMaxScore());
+        }
+
+        rowsLeft = rowsLeft - fetch;
+
+        if (rowsLeft <=0) {
+            /** Clear the scroll context on last page */
+            ClearScrollResponse clearScrollResponse = client.prepareClearScroll().addScrollId(newScrollId).get();
+            if (!clearScrollResponse.isSucceeded()) {
+                Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_SYS).increment();
+                LOG.info("Error closing the cursor context {} ", newScrollId);
+            }
+        }
+
+        cursor.setRowsLeft(rowsLeft);
+        cursor.setScrollId(newScrollId);
```

Comment:
In case of rowsLeft = 0, we still include the newScrollId in response?

Reply:
No, as part of cursor generation if `rowsleft <=0` the `cursor` is set to null, and thus omitted from response. See `DefaultCursor.generateId()`.  The code was structured like this to avoid `if...else` condition and the final logic to generate cursor id delegated to implementing class.